onnx_trt_backend.cpp

```diff
@@ -136,9 +136,34 @@ class OnnxTensorRTEvent {
   cudaEvent_t event_;
 };
 
+class CudaDeviceGuard {
+  public:
+    CudaDeviceGuard(int backend_id) {
+      if (cudaGetDevice(&current_device_) != cudaSuccess) {
+        throw std::runtime_error("Cannot run cudaGetDevice");
+      }
+      if (current_device_ != backend_id) {
+        if (cudaSetDevice(backend_id) != cudaSuccess) {
+          throw std::runtime_error("Cannot run cudaSetDevice");
+        }
+        need_restore_ = true;
+      }
+    }
+
+    ~CudaDeviceGuard() {
+      if (need_restore_) {
+        cudaSetDevice(current_device_);
+      }
+    }
+
+  private:
+    int current_device_{0};
```

Comment:
IMO, the name is a bit counter-intuitive as for the most of life-time of this object this is not current device. Maybe rename it to `previous_device_` or `saved_device_`?

Reply:
Makes sense.